### PR TITLE
#16281 adding todo items for affiliation invitation - access request

### PR DESCRIPTION
**Pull request by hfekete**
*Issue #:16281* [/bcgov/entity16281](https://github.com/bcgov/entity/issues/16281)

Adding todo items on the entity panel for open org affiliation invitations (request access type)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-filings-ui license (Apache 2.0).

### CODE REVIEW

**Comment by hfekete** [comment]
Pushing the PR first to get first feedback and answer any comments needed.
Will add TCs for added features in the next commits.

File: src/components/Dashboard/TodoList.vue
Comment (severinbeauvais) [inline src/components/Dashboard/TodoList.vue:596]: Move this to line 621-622 (and make is readonly).

File: src/components/Dashboard/TodoList.vue
Comment (severinbeauvais) [inline src/components/Dashboard/TodoList.vue:561]: Move this to "components" on line 566-567.

File: src/components/Dashboard/TodoList.vue
Comment (severinbeauvais) [inline src/components/Dashboard/TodoList.vue:876]: Can this follow the existing design of the Legal API creating any todo tasks?

Comment (hfekete) [reply src/components/Dashboard/TodoList.vue:876]: For sure, but can you give me a bit more details what should be changed and/or refer me to the specific parts of the code you want my code to be similar to?

Are you possibly referring to the loadTodoItem method and adding "Affiliation Invite" as a task first ( and to create a new FilingType) or to something else ?

Thanks!

Comment (severinbeauvais) [reply src/components/Dashboard/TodoList.vue:876]: I believe all todo items come from the Legal API. The Legal API dynamically creates todo tasks as needed (eg, an Annual Report is due).

Here, the Legal API would add a todo task if an invitation needs to be approved. The TodoList would simply display it (may need small TodoList updates to handle display).

@argush3 @vysakh-menon-aot Could one of you link to the code in Legal API where todo tasks are created?

File: tests/unit/TodoList2.spec.ts
Comment (severinbeauvais) [inline tests/unit/TodoList2.spec.ts:76]: Nice! 👍

File: tests/unit/TodoList2.spec.ts
Comment (severinbeauvais) [inline tests/unit/TodoList2.spec.ts:78]: Dunno where my previous comment went, but...

I recommend stubbing the auth service call directly instead of the axios call. (This test suite is not responsible for testing the services.)

Comment (hfekete) [reply tests/unit/TodoList2.spec.ts]: That is fair, replaced the stub from axios to auth-service.

File: src/components/Dashboard/TodoList.vue
Comment (hfekete) [reply src/components/Dashboard/TodoList.vue:876]: I am not fully familiar with all the APIs that exist as part of the system. I did work on Auth API, and currently all the data in the backend will be stored in the Auth API. 
I am assuming these two are not the same backend projects. And if so, I would prefer to avoid adding proxy backend calls.

I believe that avoiding proxy call would reduce complexity, improve maintenance and simplicity of understanding where data is coming from.

Also, I tried to put the call to load affiliation invitations at the place where data is getting loaded into the control, as my understanding is that they are not to be part of any filings or anything else other then todo reminder on that specific page.

That being said, if it is mandatory to go through Legal API to provide todo data I am not against it, just think that approach to getting data from the source API might be more straightforward.

Thank you!

Comment (hfekete) [reply src/components/Dashboard/TodoList.vue:876]: Maybe it would make sense to discuss this offline and see what is the desired approach for this specific case. 

I could get better understanding of the API relations to this frontend and their role and I can share why I chose this approach. 

I am open to advice and can update my approach/code to fit the current repo both in style and data flow.

Thanks!

Comment (severinbeauvais) [reply src/components/Dashboard/TodoList.vue:876]: I'm hoping Argus or Vysakh can provide a link for you to look at. We can discuss next week if needed.

The basic principle is that the UI should be as dumb as possible.

Comment (kialj876) [reply src/components/Dashboard/TodoList.vue:876]: hey Sev thats a good point, but I think from a backend perspective it really doesn't make sense to combine auth/lear data in the legal-api response. In general I think the apis should be as decoupled as possible and the UI can decide what it needs from each one.  I think that makes it simpler when designs change or new use cases come up and makes a lot more sense from a 3rd party api perspective too

Comment (severinbeauvais) [reply src/components/Dashboard/TodoList.vue:876]: OK, let's try to have the Todo List fetch data from both APIs...

1. do affiliation invitations have a status (eg, new)?
2. do affiliation invitations block other filings, ie, only allow 1 active filing active at a time?
3. do any other scenarios block affiliation invitations,ie, do we need `if (isAllowed(affilliation_invitation))`?

File: src/components/Dashboard/TodoList.vue
Comment (severinbeauvais) [inline src/components/Dashboard/TodoList.vue:61]: Instead of parallel conditions throughout this (already-ugly) template, could the affiliation condition be added to `showDetailsBtnBlue`?

File: src/components/Dashboard/TodoList.vue
Comment (severinbeauvais) [inline src/components/Dashboard/TodoList.vue:722]: Do you actually want to load affiliation invitations before the other tasks?

File: src/components/Dashboard/TodoList.vue
Comment (severinbeauvais) [inline src/components/Dashboard/TodoList.vue:861]: So all affiliation invitations have order=0, thus displaying above the other tasks maybe? This feels hacky,

File: src/components/Dashboard/TodoList.vue
Comment (severinbeauvais) [inline src/components/Dashboard/TodoList.vue:864]: Typo

File: src/components/Dashboard/TodoList/AffiliationInvitationDetails.vue
Comment (severinbeauvais) [inline src/components/Dashboard/TodoList/AffiliationInvitationDetails.vue:15]: Should not have comma before "will".

File: tests/unit/TodoList1.spec.ts
Comment (severinbeauvais) [inline tests/unit/TodoList1.spec.ts:258]: Is this required because of the multiple asyncs in loadData()?

Theoretically this additional wait shouldn't be needed, but if you can't figure this out quickly then I recommend leaving it and remove the await nextTick() calls (you really shouldn't need both; if you do then something's wrong with the code or test).

File: tests/unit/TodoList2.spec.ts
Comment (severinbeauvais) [inline tests/unit/TodoList2.spec.ts:140]: It would be better to assign a classname to your buttons that refers to its purpose, instead of this styling classname. I think you can just rename the existing class -- it will serve as both identifying the type of button and setting some CSS for it.

(I'm trying to avoid confusion between styling classes and "functional" classes. Styling classes should be used for styling only. Functional classes can help identify what the button is for and also include styling/CSS if needed. Another way to look at it is that "all affiliation buttons have width 45% and float-right, but not all width-45-float-right buttons are necessarily affiliation buttons".)

File: src/components/Dashboard/TodoList.vue
Comment (seeker25) [inline src/components/Dashboard/TodoList.vue:884]: This should be moved into the service itself, using await and .then afterwards seems fairly weird

File: src/components/Dashboard/TodoList.vue
Comment (severinbeauvais) [reply src/components/Dashboard/TodoList.vue:876]: OK, thanks. Resolving. (I have made other comments to make this fit more cleanly with the existing code.)

Comment (hfekete) [reply src/components/Dashboard/TodoList.vue:876]: Agree with @kialj876 
1) not that it would matter (only pending status, for specific type of the affiliation invitation is displayed)
2) does not block any filling, it is just something owner has as a todo item while invite is 'pending'
3) no scenarios that would block affiliation invitations (affiliation invitation is basically a request from someone to manage the business; access to the managing org)

File: tests/unit/TodoList1.spec.ts
Comment (hfekete) [reply tests/unit/TodoList1.spec.ts:258]: Yes, it is required because of the await in loadData. 

I believe we could do without the nextTick() but not without flushPromises now.

What I understand is that nextTick waits for next updates of the DOM for vue (which can happen multiple times) while flushPromises waits for clearing of all the promises currently in queue (including the DOM update)

OK will remove nexTick() calls

File: src/components/Dashboard/TodoList.vue
Comment (hfekete) [reply src/components/Dashboard/TodoList.vue:861]: Updated, to set order as array size. And as you previously mentioned, I moved the code to load the todos in front of the other items.

File: src/components/Dashboard/TodoList/AffiliationInvitationDetails.vue
Comment (hfekete) [reply src/components/Dashboard/TodoList/AffiliationInvitationDetails.vue:23]: Updated with new wording from UX team.

File: src/components/Dashboard/TodoList.vue
Comment (severinbeauvais) [reply src/components/Dashboard/TodoList.vue:861]: If you set order as array size then these will appear last (if the list is sorted), but if you load these before the other todos then do you actually want them first?

File: src/enums/filingStatus.ts
Comment (hfekete) [reply src/enums/filingStatus.ts:14]: Updated, will check if the invitation details `id` is set in the todo item.

File: src/components/Dashboard/TodoList.vue
Comment (hfekete) [reply src/components/Dashboard/TodoList.vue:884]: will update code a bit to be more clear

File: src/enums/filingStatus.ts
Comment (severinbeauvais) [reply src/enums/filingStatus.ts:14]: But `id` may not be unique to affiliation invitations...

Do you want me to add a value to the Filing Types enum? (AFFILIATION_INVITATION)

Comment (hfekete) [reply src/enums/filingStatus.ts:14]: I added `affiliationInvitationDetails` in the todo item:
So I am checking this: `return !!item.affiliationInvitationDetails?.id`

I wanted to avoid adding AFFILIATOIN_INVITATION as filing type because it is not a filing, like you know. Though it is requested to be displayed along filing todos. 

Is the check I am doing good enough (that part should be unique to the affiliation invitation)

File: src/components/Dashboard/TodoList.vue
Comment (hfekete) [reply src/components/Dashboard/TodoList.vue]: Moved the call to be before adding other items.

File: src/components/Dashboard/TodoList.vue
Comment (hfekete) [reply src/components/Dashboard/TodoList.vue]: I'll double check with the UX on should they be ordered first, as I do not have the explicit requirement for this page (though on the other pages affiliation invitations were always on top.)

File: src/enums/filingStatus.ts
Comment (severinbeauvais) [reply src/enums/filingStatus.ts]: Oh I see -- `id` is a property in a sub-object.

Wouldn't it be easier/cleaner to check item.name? I'm OK with adding it to the filing type because it's kind of handled like a filing/task.

Or, why not just check if item.affiliationInvitationDetails is not falsy?

File: src/components/Dashboard/TodoList.vue
Comment (severinbeauvais) [reply src/components/Dashboard/TodoList.vue]: OK, I'll leave this with you. Resolving.

File: src/enums/filingStatus.ts
Comment (hfekete) [reply src/enums/filingStatus.ts]: Yeah, it makes sense, will update to check `affiliationInvitationDetails` is not falsy

**Review by severinbeauvais — APPROVED** [review]
LGTM but please wait for more reviews before merging.

File: src/components/Dashboard/TodoList.vue
Comment (kialj876) [inline src/components/Dashboard/TodoList.vue:888]: Should this be a console.error or console.warn ?

**Review by kialj876 — APPROVED** [review]
LGTM too 👍

File: src/components/Dashboard/TodoList.vue
Comment (severinbeauvais) [reply src/components/Dashboard/TodoList.vue:888]: Or a dialog to let the user know?

Comment (hfekete) [reply src/components/Dashboard/TodoList.vue:888]: UX has no specific scenario for this, when I previously asked they said we could potentially display a generic modal for that page.

I could provide alertbox / modal saying something like following:

"Failed to fetch affiliation invite TODOs." 

Would that work ?
Or if someone can suggest example how failed API calls are handled right now ? 
I would prefer not to fail loading entire page for it though, as it might retrieve other TODO items.

Comment (severinbeauvais) [reply src/components/Dashboard/TodoList.vue:888]: The code has numerous examples of a dialog for API errors.

As this isn't a user error, I'd be ok with an alert() as a last resort, but I'd prefer to see a dialog. You probably need to create a new one as this is a specific error.

Comment (hfekete) [reply src/components/Dashboard/TodoList.vue:913]: Agreed, dialog looks much better and more user friendly.

Added preview of how the dialog looks like.
image

Also, I named it "Generic error dialog" so any error like this could reuse it by using template slots. Added title and message, and example inside the dialog component if no title and message is displayed.  
Hope that this is acceptable.

File: src/components/Dashboard/TodoList.vue
Comment (severinbeauvais) [inline src/components/Dashboard/TodoList.vue:914]: Do you ever have to reset this to false? (Example: see resetCancelPaymentErrors().)

Or does this dialog just display until the user closes it?

File: src/components/Dashboard/TodoList.vue
Comment (severinbeauvais) [reply src/components/Dashboard/TodoList.vue:913]: I think this looks great! Did you happen to check with UX folks if they like it too?

File: src/components/dialogs/GenericErrorDialog.vue
Comment (severinbeauvais) [inline src/components/dialogs/GenericErrorDialog.vue:17]: Haha, but this is the default slot so why not? 😁

File: src/components/dialogs/GenericErrorDialog.vue
Comment (severinbeauvais) [inline src/components/dialogs/GenericErrorDialog.vue:45]: You don't use this so please remove it.

File: src/components/dialogs/GenericErrorDialog.vue
Comment (severinbeauvais) [inline src/components/dialogs/GenericErrorDialog.vue:58]: I like this a lot 👍

File: src/components/dialogs/GenericErrorDialog.vue
Comment (severinbeauvais) [inline src/components/dialogs/GenericErrorDialog.vue]: Please add this back in as "generic-error-dialog" -- it can be used for unit testing.

File: src/components/Dashboard/TodoList.vue
Comment (hfekete) [reply src/components/Dashboard/TodoList.vue:906]: OK, apparently this is old design. I updated it to match new design. Also changed it a bit so that it should always conform to their style (e.g. can change the icon, but it is not a slot anymore, just icon mdi font, etc ...)

New design looks like this:
image

File: src/components/Dashboard/TodoList.vue
Comment (hfekete) [reply src/components/Dashboard/TodoList.vue:907]: I do this on line ~13 `@okay="fetchAffiliationInvitationsErrorDialog=false"`

There is no cleanup actions other then closing the dialog, so wanted to avoid adding unnecessary method. If need for method arises, it can be added at that time.

File: src/components/Dashboard/TodoList.vue
Comment (severinbeauvais) [reply src/components/Dashboard/TodoList.vue:906]: Looks good to me.

Kial, OK to resolve this conversation?

**Comment by severinbeauvais** [comment]
LGTM.

Travis, Karim, any comments?

**Review by JazzarKarim — APPROVED** [review]
LGTM 👍

**Pull request merged**
